Updates to Runtime Macros Article (#631)
substrate-developer-hub/substrate-developer-hub.github.io: docs/knowledgebase/runtime/macros.md

## Introduction

> Within Substrate Runtime Macros engineers are given the opportunity to focus on runtime logic rather

Reviewer: ```suggestion
Within a Substrate runtime, macros allow engineers to focus on runtime logic rather
```

**When to Use**

> To define storage items in a runtime palette. A storage item definition includes:

Reviewer: ```suggestion
To define storage items in a pallet. A storage item definition includes:
```

Reviewer 2: btw, is there any situation to use `FRAME pallet` better than `runtime pallet`? Or they could be used interchangeably most of the time?

Reviewer 3: This was part of a **long long and intense** discussion. I still find it okay to say runtime palette but the consensus was to **stay with FRAME palette**. A palette is always part of the runtime, so `runtime palette` is redundant

**What It Does**

This macro takes a succinct statement of:

```
#vis #name get(fn #getter) config(#field_name) build(#closure): #type = #default;
```

Replacing it with a new struct type defined by `#name` and have it implement the data type storage
trait.

The macro also declares and implements `Store` traits to set up the pallet to have storage space, and
implements getters on the `Module` struct defined by `decl_module!`.

**Docs and Notes**

- [API Documentation](https://substrate.dev/rustdocs/v2.0.0-rc6/frame_support/macro.decl_storage.html)
- `Store` trait is declared with each storage name, thus becoming an associated type inside the
  trait.
- `Module` implementation contains the getter function and metadata of each storage item.
- Each of the storage items becomes a struct. Above in the expanded code `Something` struct type

>   [is declared and implements `StorageValue<u32>` trait](https://gist.github.com/jimmychu0807/c4a88ec8e0342ee9f4e14bd26287324e#file-pallet-template-expanded-rs-L144-L164)

Reviewer 3: We need to find a better way than pointing to personal gists.

Author: Yes, I agree. Is the Gist here potentially out-of-date since it may be an expanded version of `rc5` or something? We would need a strategy around making sure this stays up-to-date.

Reviewer 3: Yeah, for sure. I don't think that it contains updated code.

Replacing it with a new struct type defined by `#name` and have it implement the data type storage
trait.

> The macro also declares and implements `Store` traits to set up the pallet to have storage space, and

Reviewer: ```suggestion
The macro also declares and implements the `Store` trait to set up the pallet to have storage space, and
```

**What It Does**

This macro defines pallet events by implementing `Event` enum type, with each event type in the
macro being an enum variant inside `Event`.

**Docs and Notes**

- [API Documentation](https://substrate.dev/rustdocs/v2.0.0-rc6/frame_support/macro.decl_event.html)
- `Event` or `Event<T>` type is defined, assigned to `RawEvent`.
- `RawEvent<...all generic types used>` enum type is defined with all events specified in the macro
  as its enum variants.
- The macro implements various helper traits for`RawEvent`, including data encoding/decoding,
  `core::cmp::PartialEq`, `core::cmp::Clone`, and `core::fmt::Debug`.
- `RawEvent<...>` implements `metadata()` function to retrieve meta-data for events emitted in the
  pallet.

### decl_error!

**When to Use**

To define error types which a pallet may return in its dispatchable functions. Dispatchable
functions return `DispatchResult`, with either an `Ok(())`, or `DispatchError` containing custom
errors defined in this macro.

**What It Does**

This macro defines `Error<T: Trait>` struct type, and implement helper methods for mapping each
error type to sequential error code and corresponding string.

> One key is that the macro automatically implements the `From<Error<T>>` trait for `DispatchError`.

Reviewer 3: Kinda ambiguous. One key aspect/feature ??

  encoding/decoding traits implemented for the enum. Various conversion traits `Event` also
  implements `TryInto<pallets::Event<Runtime>>` trait to extract the event out from the enum type.
- `Origin` enum type is defined with helper traits, e.g. `PartialEq`, `Clone`, `Debug` implemented.
  This enum type defines who calls an extrinsic, `NONE`, `ROOT`, or signed by a particular account.

>   The origin can also be defined by other pallets, not only systems.

Reviewer: ```suggestion
  The origin can also be defined by other pallets, not only frame system.
```

Reviewer: "Frame system" is the name of a pallet.
